zetaclient/chains/bitcoin/observer/event.go

```diff
@@ -0,0 +1,233 @@
+package observer
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"math/big"
+
+	cosmosmath "cosmossdk.io/math"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/pkg/errors"
+
+	"github.com/zeta-chain/node/pkg/chains"
+	"github.com/zeta-chain/node/pkg/coin"
+	"github.com/zeta-chain/node/pkg/constant"
+	"github.com/zeta-chain/node/pkg/crypto"
+	"github.com/zeta-chain/node/pkg/memo"
+	crosschaintypes "github.com/zeta-chain/node/x/crosschain/types"
+	"github.com/zeta-chain/node/zetaclient/compliance"
+	"github.com/zeta-chain/node/zetaclient/config"
+	"github.com/zeta-chain/node/zetaclient/logs"
+)
+
+// InboundProcessability is an enum representing the processability of an inbound
+type InboundProcessability int
+
+const (
+	// InboundProcessabilityGood represents a processable inbound
+	InboundProcessabilityGood InboundProcessability = iota
+
+	// InboundProcessabilityDonation represents a donation inbound
+	InboundProcessabilityDonation
+
+	// InboundProcessabilityComplianceViolation represents a compliance violation
+	InboundProcessabilityComplianceViolation
+)
+
+// BTCInboundEvent represents an incoming transaction event
+type BTCInboundEvent struct {
+	// FromAddress is the first input address
+	FromAddress string
+
+	// ToAddress is the ZEVM receiver address
+	ToAddress string
+
+	// Value is the amount of BTC
+	Value float64
+
+	// DepositorFee is the deposit fee
+	DepositorFee float64
+
+	// MemoBytes is the memo of inbound
+	MemoBytes []byte
+
+	// MemoStd is the standard inbound memo if it can be decoded
+	MemoStd *memo.InboundMemo
+
+	// BlockNumber is the block number of the inbound
+	BlockNumber uint64
+
+	// TxHash is the hash of the inbound
+	TxHash string
+}
+
+// IsProcessable checks if the inbound event is processable
+func (event *BTCInboundEvent) CheckProcessability() InboundProcessability {
+	// compliance check on sender and receiver addresses
+	if config.ContainRestrictedAddress(event.FromAddress, event.ToAddress) {
+		return InboundProcessabilityComplianceViolation
+	}
+
+	// compliance check on receiver, revert/abort addresses in standard memo
+	if event.MemoStd != nil {
+		if config.ContainRestrictedAddress(
+			event.MemoStd.Receiver.Hex(),
+			event.MemoStd.RevertOptions.RevertAddress,
+			event.MemoStd.RevertOptions.AbortAddress,
+		) {
+			return InboundProcessabilityComplianceViolation
+		}
+	}
+
+	// donation check
+	if bytes.Equal(event.MemoBytes, []byte(constant.DonationMessage)) {
+		return InboundProcessabilityDonation
+	}
+
+	return InboundProcessabilityGood
+}
+
+// DecodeEventMemoBytes decodes the memo bytes as either standard or legacy memo
+func (ob *Observer) DecodeEventMemoBytes(event *BTCInboundEvent) error {
+	var (
+		err      error
+		memoStd  *memo.InboundMemo
+		receiver ethcommon.Address
+	)
+
+	// skip decoding donation tx as it won't go through zetacore
+	if bytes.Equal(event.MemoBytes, []byte(constant.DonationMessage)) {
```

Comment:
We shouldn't need this call here if we already have this check in `Processability`

Reply:
It's a good point and it's by intention. Here is the workflow:
step-1: call `DecodeMemoBytes` so that we know the `receiver`, `revert address`, `abort address` etc.
step-2: call `CheckEventProcessability` against decoded addresses to not violate the compliance.


If the early on `DonationMessage` gets removed, the followed logic will try decoding it as a legacy memo and then fails because the message `I am rich!` is too short to be valid(a receiver address required and memo must be` >= 20 bytes`

```
// decode event memo bytes
err := event.DecodeMemoBytes(ob.Chain().ChainId)
if err != nil {
	return nil
}

// check if the event is processable
if !ob.CheckEventProcessability(event) {
	return nil
}
```

Reply:
Why would `DonationMessage` gets removed, on which step?

Reply:
[removed on this step](https://github.com/zeta-chain/node/blob/feat-btc-revert-address/zetaclient/chains/bitcoin/observer/event.go#L173)